testing/e2e/infra/provision.go

```diff
@@ -30,6 +31,18 @@ func (i *infra) Provision(ctx context.Context, tenantId, subscriptionId string)
 		TenantId:       tenantId,
 	}
 
+	if i.ServicePrincipal != nil {
+		if applicationObjectId == "" {
+			return ret, logger.Error(lgr, fmt.Errorf("application object id must be provided when provisioning infrastructure with service principal options"))
+		}
+		credName := i.Name + "-cred"
+		spOpt, err := clients.GetServicePrincipalOptions(ctx, applicationObjectId, credName)
+		if err != nil {
+			return ret, logger.Error(lgr, fmt.Errorf("getting app with credential: %w", err))
+		}
+		i.ServicePrincipal = spOpt
```

Comment:
it's a weird pattern that we check if `i.ServicePrincipal != nil` when entering this block then overwrite it. Maybe the field on infra should be called useSp or something?

Reply:
good catch, went with an enum to make it clearer